client/servers/manager.go

```diff
@@ -0,0 +1,529 @@
+}
+
+// Pinger is an interface for pinging a server to see if it is healthy.
+type Pinger interface {
+	Ping(addr net.Addr) (bool, error)
+}
+
+// serverList is a local copy of the struct used to maintain the list of
+// Nomad servers used by Manager.
+//
+// NOTE(sean@): We are explicitly relying on the fact that serverList will
+// be copied onto the stack.  Please keep this structure light.
+type serverList struct {
+	// servers tracks the locally known servers.  List membership is
+	// maintained by Serf.
+	servers []*Server
+}
+
+type Manager struct {
+	// listValue manages the atomic load/store of a Manager's serverList
+	listValue atomic.Value
+	listLock  sync.Mutex
+
+	// rebalanceTimer controls the duration of the rebalance interval
+	rebalanceTimer *time.Timer
+
+	// shutdownCh is a copy of the channel in Nomad.Client
+	shutdownCh chan struct{}
+
+	logger *log.Logger
+
+	// numNodes is used to estimate the approximate number of nodes in
+	// a cluster and limit the rate at which it rebalances server
+	// connections. This should be read and set using atomic.
+	numNodes int32
+
+	// connPoolPinger is used to test the health of a server in the connection
+	// pool. Pinger is an interface that wraps client.ConnPool.
+	connPoolPinger Pinger
+
+	// notifyFailedBarrier is acts as a barrier to prevent queuing behind
+	// serverListLog and acts as a TryLock().
+	notifyFailedBarrier int32
+
+	// offline is used to indicate that there are no servers, or that all
+	// known servers have failed the ping test.
+	offline int32
+}
+
+// New is the only way to safely create a new Manager struct.
+func New(logger *log.Logger, shutdownCh chan struct{}, connPoolPinger Pinger) (m *Manager) {
+	m = new(Manager)
+	m.logger = logger
+	m.connPoolPinger = connPoolPinger // can't pass *Nomad.ConnPool: import cycle
+	m.rebalanceTimer = time.NewTimer(clientRPCMinReuseDuration)
+	m.shutdownCh = shutdownCh
+	atomic.StoreInt32(&m.offline, 1)
+
+	l := serverList{}
+	l.servers = make([]*Server, 0)
+	m.saveServerList(l)
+	return m
+}
+
+// Start is used to start and manage the task of automatically shuffling and
+// rebalancing the list of Nomad servers.  This maintenance only happens
+// periodically based on the expiration of the timer.  Failed servers are
+// automatically cycled to the end of the list.  New servers are appended to
+// the list.  The order of the server list must be shuffled periodically to
+// distribute load across all known and available Nomad servers.
+func (m *Manager) Start() {
+	for {
+		select {
+		case <-m.rebalanceTimer.C:
+			m.RebalanceServers()
+			m.refreshServerRebalanceTimer()
+
+		case <-m.shutdownCh:
+			m.logger.Printf("[INFO] manager: shutting down")
+			return
+		}
+	}
+}
+
+func (m *Manager) SetServers(servers []*Server) {
+	// Hot Path
+	if len(servers) == 0 {
+		return
+	}
+
+	m.listLock.Lock()
+	defer m.listLock.Unlock()
+
+	// Update the server list
+	l := m.getServerList()
+	l.servers = servers
```

Comment:
There's no need to get the existing server list just to override it. These lines can be removed and saveServerList can  do:

```go
m.saveServerList(serverList{servers})
```